src/cryptography/nostr_bech32.cpp

```diff
@@ -0,0 +1,743 @@
+
+    return 0;
+}
+
+static int parse_nostr_bech32_type(const char *prefix, enum nostr_bech32_type *type) {
+    // Parse type
+    if (strcmp(prefix, "note") == 0) {
+        *type = NOSTR_BECH32_NOTE;
+        return 1;
+    } else if (strcmp(prefix, "npub") == 0) {
+        *type = NOSTR_BECH32_NPUB;
+        return 1;
+    } else if (strcmp(prefix, "nsec") == 0) {
+        *type = NOSTR_BECH32_NSEC;
+        return 1;
+    } else if (strcmp(prefix, "nprofile") == 0) {
+        *type = NOSTR_BECH32_NPROFILE;
+        return 1;
+    } else if (strcmp(prefix, "nevent") == 0) {
+        *type = NOSTR_BECH32_NEVENT;
+        return 1;
+    } else if (strcmp(prefix, "nrelay") == 0) {
+        *type = NOSTR_BECH32_NRELAY;
+        return 1;
+    } else if (strcmp(prefix, "naddr") == 0) {
+        *type = NOSTR_BECH32_NADDR;
+        return 1;
+    }
+
+    return 0;
+}
+
+static int parse_nostr_bech32_note(struct cursor *cur, struct bech32_note *note) {
+    return pull_bytes(cur, 32, &note->event_id);
+}
+
+static int parse_nostr_bech32_npub(struct cursor *cur, struct bech32_npub *npub) {
+    return pull_bytes(cur, 32, &npub->pubkey);
+}
+
+static int parse_nostr_bech32_nsec(struct cursor *cur, struct bech32_nsec *nsec) {
+    return pull_bytes(cur, 32, &nsec->privkey);
+}
+
+static int tlvs_to_relays(struct nostr_tlvs *tlvs, struct relays *relays) {
+    struct nostr_tlv *tlv;
+    struct str_block *str;
+
+    relays->num_relays = 0;
+
+    for (int i = 0; i < tlvs->num_tlvs; i++) {
+        tlv = &tlvs->tlvs[i];
+        if (tlv->type != TLV_RELAY)
+            continue;
+
+        if (relays->num_relays + 1 > MAX_RELAYS)
+            break;
+
+        str = &relays->relays[relays->num_relays++];
+        str->start = (const char*)tlv->value;
+        str->end = (const char*)(tlv->value + tlv->len);
+    }
+
+    return 1;
+}
+
+static uint32_t decode_tlv_u32(const uint8_t *bytes) {
+    uint32_t *be32_bytes = (uint32_t*)bytes;
+    return htobe32(*be32_bytes);
+}
+
+static void encode_u32_tlv(const uint32_t *value, uint8_t *be32_value) {
+    uint32_t be32 = htobe32(*value);
+    memcpy(be32_value, (uint8_t*)&be32, 4);
+}
+
+static int parse_nostr_bech32_nevent(struct cursor *cur, struct bech32_nevent *nevent) {
+    struct nostr_tlvs tlvs;
+    struct nostr_tlv *tlv;
+
+    if (!parse_nostr_tlvs(cur, &tlvs))
+        return 0;
+
+    if (!find_tlv(&tlvs, TLV_SPECIAL, &tlv))
+        return 0;
+
+    if (tlv->len != 32)
+        return 0;
+
+    nevent->event_id = tlv->value;
+
+    if (find_tlv(&tlvs, TLV_AUTHOR, &tlv)) {
+        nevent->pubkey = tlv->value;
+    } else {
+        nevent->pubkey = NULL;
+    }
+
+    if(find_tlv(&tlvs, TLV_KIND, &tlv)) {
+        nevent->kind = decode_tlv_u32(tlv->value);
+        nevent->has_kind = true;
+    } else {
+        nevent->has_kind = false;
+    }
+
+    return tlvs_to_relays(&tlvs, &nevent->relays);
+}
+
+static int parse_nostr_bech32_naddr(struct cursor *cur, struct bech32_naddr *naddr) {
```

Comment:
Thoughts for the future:

An idiomatic C++ wrapper for this functionality would escalate failing error codes returned from this function into thrown exceptions, to be handled as calling code deems fit.